waku/v2/protocol/waku_filter/protocol.nim

```diff
@@ -0,0 +1,282 @@
+  ## check for elapsed time and if it's been too long, remove the peer.
+  for sub in subs:
+    let subKey: string = $(sub)
+    
+    if not wf.failedPeers.hasKey(subKey):
+      # add the peer to the failed peers table.
+      wf.failedPeers[subKey] = Moment.now() 
+      return
+
+    let elapsedTime = Moment.now() - wf.failedPeers[subKey]
+    if elapsedTime > wf.timeout:
+      wf.failedPeers.del(subKey)
+
+      let index = wf.subscriptions.find(sub)
+      wf.subscriptions.delete(index)
+
+
+proc handleMessage*(wf: WakuFilter, pubsubTopic: string, msg: WakuMessage) {.async.} =
+  if wf.subscriptions.len <= 0:
+    return
+
+  var failedSubscriptions: seq[Subscription]
+  var connectedSubscriptions: seq[Subscription]
+
+  for sub in wf.subscriptions:
+    # TODO: Review when pubsubTopic can be empty and if it is a valid case
+    if sub.pubSubTopic != "" and sub.pubSubTopic != pubsubTopic:
+      continue
+
+    if msg.contentTopic notin sub.contentTopics:
+      continue
+
+    let rpc = FilterRPC(
+      requestId: sub.requestId,
+      push: MessagePush(messages: @[msg])
+    )
+
+    let res = await wf.sendFilterRpcToPeer(rpc, sub.peer)
+    if res.isErr():
+      waku_filter_errors.inc(labelValues = [res.error()])
+      failedSubscriptions.add(sub)
+      continue
+      
+    connectedSubscriptions.add(sub)
+
+  wf.removePeerFromFailedPeersTable(connectedSubscriptions)
+
+  wf.handleClientError(failedSubscriptions)
+
+
+### Send subscription/unsubscription
+
+proc subscribe(wf: WakuFilter, pubsubTopic: string, contentTopics: seq[ContentTopic], peer: RemotePeerInfo): Future[WakuFilterResult[string]] {.async, gcsafe.} =
+  let id = generateRequestId(wf.rng)
+  let rpc = FilterRPC(
+    requestId: id,
+    request: FilterRequest(
+      subscribe: true,
+      pubSubTopic: pubsubTopic,
+      contentFilters: contentTopics.mapIt(ContentFilter(contentTopic: it))
+    )
+  )
+
+  let res = await wf.sendFilterRpcToRemotePeer(rpc, peer)
+  if res.isErr():
+    waku_filter_errors.inc(labelValues = [res.error()])
+    return err(res.error())
+    
+  return ok(id)
+
+proc subscribe*(wf: WakuFilter, pubsubTopic: string, contentTopics: seq[ContentTopic]): Future[WakuFilterResult[string]] {.async, gcsafe.} =
+  let peerOpt = wf.peerManager.selectPeer(WakuFilterCodec)
+  if peerOpt.isNone():
+    waku_filter_errors.inc(labelValues = [peerNotFoundFailure])
+    return err(peerNotFoundFailure)
+
+  return await wf.subscribe(pubsubTopic, contentTopics, peerOpt.get())
+
+
+proc unsubscribe(wf: WakuFilter, pubsubTopic: string, contentTopics: seq[ContentTopic], peer: RemotePeerInfo): Future[WakuFilterResult[void]] {.async, gcsafe.} =
+  let id = generateRequestId(wf.rng)
+  let rpc = FilterRPC(
+    requestId: id, 
+    request: FilterRequest(
+      subscribe: false,
+      pubSubTopic: pubsubTopic,
+      contentFilters: contentTopics.mapIt(ContentFilter(contentTopic: it))
+    )
+  )
+
+  let res = await wf.sendFilterRpcToRemotePeer(rpc, peer)

```

Comment:
See comment above about using `sendFilterRpcToPeer()`

Reply:
Please, check my answer to your other comment :)